catalog/core/catalog-core-subscriptionstore/subscriptionstore-impl/src/main/java/org/codice/ddf/catalog/subscriptionstore/SubscriptionContainerImpl.java

```diff
@@ -0,0 +1,396 @@
+    }
+
+    /**
+     * Ensure the identifier points to a valid subscription, and that it exists in the container.
+     *
+     * @param operation customize the message based on where this validation is occurring.
+     */
+    private void validateContainment(SubscriptionIdentifier identifier, String operation) {
+        if (get(identifier) == null) {
+            LOGGER.debug("Target for subscription {} [ {} | {} ] does not exist",
+                    operation,
+                    identifier.getId(),
+                    identifier.getTypeName());
+            throw new SubscriptionStoreException(format("Subscription [%s] does not exist. ",
+                    identifier.getId()));
+        }
+    }
+
+    private void validateSubscription(Subscription subscription) {
+        notNull(subscription, "Subscription object cannot be null. ");
+    }
+
+    private void validateIdentifier(SubscriptionIdentifier identifier) {
+        notNull(identifier, "Subscription identifier cannot be null. ");
+        notEmpty(identifier.getId(), "Subscription ID string cannot be null or empty. ");
+        notEmpty(identifier.getTypeName(), "Subscription type string cannot be null or empty. ");
+
+        // TODO: Get Approval - Backslash is not the ECQL escape character, thus we need to forbid single quotes
+        if (identifier.getId()
+                .contains(SINGLE_QUOTE)) {
+            throw new SubscriptionStoreException("Subscription ID contains invalid characters");
+        }
+    }
+
+    private void validateType(SubscriptionType type) {
+        notNull(type, "Subscription type cannot be null. ");
+        notEmpty(type.getTypeName(), "Subscription type string cannot be null or empty. ");
+    }
+
+    private void validateSerialized(MarshalledSubscription marshalledSubscription) {
+        notNull(marshalledSubscription, "Serialized subscription cannot be null. ");
+        notEmpty(marshalledSubscription.getFilter(),
+                "Serialized filter string cannot be null or empty. ");
+        notEmpty(marshalledSubscription.getCallbackAddress(),
+                "Callback address cannot be null or empty. ");
+    }
+
+    /**
+     * Configuration for the javax.cache (For now, Cache2K is used).
+     */
+    @SuppressWarnings("unchecked")
+    Cache<String, CachedSubscription> createCache(SubscriptionCacheLoader cacheLoader,
+            SubscriptionCacheWriter cacheWriter) {
+        CachingProvider cachingProvider =
+                Caching.getCachingProvider(JCacheProvider.class.getClassLoader());
+        CacheManager cacheManager = cachingProvider.getCacheManager();
+
+        MutableConfiguration<String, CachedSubscription> config =
+                cleanMutableTypedConfig().setExpiryPolicyFactory(EternalExpiryPolicy::new)
+                        .setCacheLoaderFactory(() -> cacheLoader)
+                        .setCacheWriterFactory(() -> cacheWriter)
+                        .setStoreByValue(false)
```

Comment:
This bundle only works using store-by-reference unless we can make osgi internals serializable. I would argue it only _makes sense_ using store-by-reference because dynamically registered services shut-down with the system and need to be re-registered anyway.